nusight2/src/client/components/three/three_fiber.tsx

```diff
@@ -44,3 +43,33 @@ export const PerspectiveCamera = ({
   }, [lookAt]);
   return <perspectiveCamera ref={ref} {...props} />;
 };
+
+export const OrthographicCamera = (
+  props: JSX.IntrinsicElements["orthographicCamera"] & {
+    /** Making it manual will stop responsiveness, and you have to calculate aspect ratio yourself. */
+    manual?: boolean;
+  },
+) => {
+  const ref = React.useRef<THREE.OrthographicCamera>(null);
+  const three = useThree();
+  React.useEffect(() => {
+    const camera = ref.current;
+    if (camera) {
+      three.set({ camera });
+      three.setSize(three.gl.domElement.clientWidth, three.gl.domElement.clientHeight);
+    }
+  }, []);
+  return <orthographicCamera ref={ref} {...props} />;
+};
+
+export const RawShaderMaterial = (
+  props: JSX.IntrinsicElements["rawShaderMaterial"] & {
+    uniforms: { [k: string]: any };
+  },
+) => {
+  const uniforms = React.useMemo(() => props.uniforms, []);
+  for (const key in props.uniforms) {
+    uniforms[key].value = props.uniforms[key].value;
+  }
+  return <rawShaderMaterial {...props} uniforms={uniforms} />;
+};
```

Comment:
Does this reuse the same shader material for the same uniforms?
Shouldn't you keep using a clone of the same shader and update the uniforms?
Or does three fibre handle not recompiling the shader each time.

Reply:
tl;dr Yes it reuses the compiled shaders. No need to clone.

R3F does nothing special, all it does is match the lifetime of rendered objects to the lifetime of the component.

If we ignore uniforms, this component is essentially equivalent to:

```ts
export const RawShaderMaterial = (props) => {
  const shader = React.useMemo(() => new THREE.RawShaderMaterial(), []);
  shader.vertexShader = props.vertexShader;
  shader.fragmentShader = props.fragmentShader;
  return <primitive object={shader}/>;
}
```

i.e. all it does it cache the instance.

So your real question is, does three.js recompile the shaders every time we render?

And the answer is no, it has an internal [WebGLShaderCache](https://github.com/mrdoob/three.js/blob/master/src/renderers/webgl/WebGLShaderCache.js), which thanks to string interning, will always hit, given the same string:

https://github.com/mrdoob/three.js/blob/c02fb47ade2a3666b9b8845cb159dcbe3d7bc932/src/renderers/webgl/WebGLShaderCache.js#L96

However three.js does not like you replacing `shader.uniforms` with a new object after first-render, which is why I cache the uniforms object itself and instead mutate individual values. I'll see if I can write a test for it.